File: app/src/main/java/io/neoterm/setup/connection/OfflineConnection.java
package io.neoterm.setup.connection;

import java.io.IOException;
import java.io.InputStream;

import io.neoterm.setup.SourceConnection;

/**
 * @author kiva
 */

public abstract class OfflineConnection implements SourceConnection {
    private InputStream inputStream;

    protected abstract InputStream openInputStream() throws IOException;

    @Override
    public InputStream getInputStream() throws IOException {
        if (inputStream == null) {
            inputStream = openInputStream();
        }
        return inputStream;
    }

    @Override
    public int getSize() {
        if (inputStream != null) {
            try {
                return inputStream.available();
            } catch (IOException e) {
                e.printStackTrace();
                return 0;
            }
        }
        return 0;
    }

    @Override
    public void close() {
        if (inputStream != null) {
            try {
                inputStream.close();
            } catch (IOException ignore) {
                ignore.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/io/neoterm/setup/connections/OfflineConnection.kt
package io.neoterm.setup.connections

import io.neoterm.setup.SourceConnection
import java.io.IOException
import java.io.InputStream

/**
 * @author kiva
 */

abstract class OfflineConnection : SourceConnection {
    private var inputStream: InputStream? = null

    @Throws(IOException::class)
    protected abstract fun openInputStream(): InputStream

    @Throws(IOException::class)
    override fun getInputStream(): InputStream {
        if (inputStream == null) {
            inputStream = openInputStream()
        }
        return inputStream!!
    }

    override fun getSize(): Int {
        if (inputStream != null) {
            try {
                return inputStream!!.available()
            } catch (e: IOException) {
                e.printStackTrace()
                return 0
            }

        }
        return 0
    }

    override fun close() {
        if (inputStream != null) {
            try {
                inputStream!!.close()
            } catch (ignore: IOException) {
                ignore.printStackTrace()
            }

        }
    }
}
